### PR TITLE
feat(bottlecap): add Step Functions trace extraction

### PR DESCRIPTION
**Pull request by duncanista**
# What?

Allows Step Function invokes to Lambda to appear in the same trace

Screenshot 2024-11-12 at 4 24 43 PM
Screenshot 2024-11-12 at 4 25 03 PM

# How

Generating trace context data through deterministic methods

# Notes

- Step Functions don't need an inferred span, so we remove it in case a generated_span_context exists

### CODE REVIEW

File: bottlecap/src/lifecycle/invocation/processor.rs [hunk 1]
```diff
@@ -23,7 +23,13 @@ use crate::{
     telemetry::events::{ReportMetrics, Status},
     traces::{
         context::SpanContext,
-        propagation::{DatadogCompositePropagator, Propagator},
+        propagation::{
+            text_map_propagator::{
+                DatadogHeaderPropagator, DATADOG_PARENT_ID_KEY, DATADOG_SPAN_ID_KEY,
+                DATADOG_TRACE_ID_KEY,
+            },
+            DatadogCompositePropagator, Propagator,
```
Comment (purple4reina) [inline bottlecap/src/lifecycle/invocation/processor.rs:31]: At point in the invocation is this happening? Is this extracting context from the inbound event? Or is it getting it from the start invocation request headers?

Just asking cuz it doesn't look like we have support for w3c headers. Is that correct?

Comment (duncanista) [reply bottlecap/src/lifecycle/invocation/processor.rs:31]: I'm not sure which line you're talking about, you left the comment in the import

Comment (duncanista) [reply bottlecap/src/lifecycle/invocation/processor.rs:31]: > Just asking cuz it doesn't look like we have support for w3c headers. Is that correct?
We do

Comment (purple4reina) [reply bottlecap/src/lifecycle/invocation/processor.rs:31]: Ahhh, didn't realize this is just the import. Thanks for answering the question!

File: bottlecap/src/lifecycle/invocation/triggers/step_function_event.rs [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+
+    fn get_arn(&self, _region: &str) -> String {
+        if let Some(sm) = &self.state_machine {
+            return sm.id.clone();
+        }
+
+        String::new()
+    }
+
+    fn get_carrier(&self) -> HashMap<String, String> {
```
Comment (duncanista) [inline bottlecap/src/lifecycle/invocation/triggers/step_function_event.rs:102]: No carrier, we generate the span context. The decision behind this is that the customer might be using W3C or other style, and we cannot just generate a carrier based on that, its easier to just return a `SpanContext`

File: bottlecap/src/lifecycle/invocation/triggers/step_function_event.rs [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+                "1914fe7789eb32be".to_string(),
+            )]),
+            links: vec![],
+        };
+
+        assert_eq!(span_context, expected);
+    }
+
+    #[test]
+    fn test_generate_parent_id() {
```
Comment (duncanista) [inline bottlecap/src/lifecycle/invocation/triggers/step_function_event.rs:326]: These tests are the ones from the Go Agent, so we can see that data is generated as expected.

**Review by agocs — APPROVED** [review]
LGTM

**Pull request merged**